docs/adrs/010-logs.md

```diff
@@ -0,0 +1,35 @@
+# ADR 010: What to do with logs?
+
+**Problem:** when commands fail, you usually want to look at the process logs to see what went wrong.
+You also might want to see logs as the system is building.
+This is more complicated than just "ok, so stream them" because we plan to add remote builders.
+How do we persist logs so you can look at them later, for example when a CI job fails?
+
+It seems like the best solution would be to have both a log store and a log sink.
+Specifically:
+
+1. during evaluation of jobs, logs can be streamed.
+   This would let us show them in the CLI or in a web UI.
+2. Logs could be stored like job artifacts in the content-addressable store.
+
+Part 1 can be implemented today, but part two needs a little bit more groundwork.
+Specifically, how do you keep track of which logs came from which jobs?
+You could store them with the artifacts, but that seems unreasonable in the case of failing jobs (since it would be unreasonable to store artifacts for a failing job.)
+
+Instead, it seems like we need the concept of an invocation, which could keep track of which jobs it did and didn't build, along with log outputs.
```

Comment:
Invocations seem to more be answering the question of handling longs when you think across runs. It still seems that just the job hash is all that is needed to tell which logs come from which job.

Reply:
I'm not 100% sure what you mean here, actually. I suppose that it might be the case that logs are attached to a job, but where do we keep track of that? The invocation seems like a reasonable way to go; do you see a better place?

Reply:
> Specifically, how do you keep track of which logs came from which jobs?

Currently we generate a spot to store artifacts for each job. I don't see why you couldn't just create a spot next to/with the same hash key to store the logs. If the jobs hash is `1234` and I store the artifacts in `jobs/1234/artifacts`, couldn't you just store the logs in `jobs/1234/logs`?

If the issue here is that you want to keep around logs from old job runs, an invocation hash or just timestamp sounds necessary, but I don't see how an invocation hash is really need to answer the original question.

Reply:
> I don't see why you couldn't just create a spot next to/with the same hash key to store the logs.

Two reasons:

1. we store artifacts in a content-addressable store. Logs frequently have timestamps in them. That's good! But it also means that then our artifacts wouldn't be reproducible.
  - this means I'm a little leery of putting them in the CAS *at all* but it's also fine? I expect we'll have to make some sort of garbage collector for the CAS at some point regardless of this decision. If anything, this'll just speed that up.
3. it wouldn't always be reasonable to store artifacts for a failed build (I'm not sure if it'd ever be, actually?) But it'd be always-reasonable to store logs for a failed build, for debugging after the fact.

Basically: they do different jobs and have slightly different storage needs.